src/rewrite.rs

```diff
@@ -24,10 +26,19 @@ impl<T: Rewrite> Rewrite for ptr::P<T> {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub(crate) struct QueryId {
+    pub(crate) shape: Shape,
+    pub(crate) span: SpanData,
+}
+
+pub(crate) type Memoize = Option<HashMap<QueryId, Option<String>>>;
+
 #[derive(Clone)]
 pub(crate) struct RewriteContext<'a> {
     pub(crate) parse_sess: &'a ParseSess,
     pub(crate) config: &'a Config,
+    pub(crate) memoize: Rc<RefCell<Memoize>>,
```

Comment:
Can you use `Cell` instead of `RefCell`, similar to other fields here?

Reply:
It isn't possible to mutate things behind `Cell` in-place (insert to hashmap), I think.

Reply:
You can `take`, mutate and replace. More ceremony but I believe more efficient after optimizations.

Reply:
Hmm, so basically `RefCell<T>` is equivalent to `Cell<Option<T>>` with unwraping on `borrow_mut` and replacing on drop? Interesting. Can you explain why it would optimize better? About similarity to other fields, there is a `pub(crate) skipped_range: Rc<RefCell<Vec<(usize, usize)>>>` there as well.

Is it possible/better to remove all cells from rewrite context, pass it by `&mut` to most functions that currently get it by `&`, and wrap it in a `Rc<RefCell<` if/when we need to clone it? Certainly out of scope for this PR.

Reply:
> Can you explain why it would optimize better?

RefCell enforces borrow checking rules at runtime which has a small cost.

Reply:
Replaced with Cell